### PR TITLE
Metrics Filter Dropdown

### PR DESCRIPTION
**Pull request by jbrooksuk**
This is a proof of concept whilst I flesh out some ideas I have with it.

### CODE REVIEW

**Comment by jbrooksuk** [comment]
Also, it's all inline for now because it's quicker to test with.

**Comment by jbrooksuk** [comment]
Grouping the points in JS will be quite slow, especially when there are thousands of them, so I propose we add three new API endpoints:

- `/api/v1/metrics/{metricid}/points/hourly`
- `/api/v1/metrics/{metricid}/points/daily`
- `/api/v1/metrics/{metricid}/points/weekly`

Then do the grouping within the routes.

CC @cachethq/owners

**Comment by jbrooksuk** [comment]
This is quite complex actually, since we also need to be accounting for sum and averages too, so the metric system itself may need a small rewrite.

**Comment by GrahamCampbell** [comment]
Don't think we should add separate endpoints.

**Comment by GrahamCampbell** [comment]
Look what cloudflare do for analytics, and sums.

File: app/Repositories/MetricRepository.php [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+use CachetHQ\Cachet\Models\Metric;
+use DateInterval;
+use Illuminate\Support\Facades\Config;
+use Illuminate\Support\Facades\DB;
+use Jenssegers\Date\Date;
+
+class MetricRepository
+{
+    /**
+     * @var string
```
Comment (GrahamCampbell) [inline app/Repositories/MetricRepository.php]: missing description

File: app/Repositories/MetricRepository.php [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+     */
+    protected function getPointsByHour(Metric $metric, $hour)
+    {
+        $dateTime = (new Date())->setTimezone($this->dateTimeZone);
+        $dateTime->sub(new DateInterval('PT'.$hour.'H'));
+        $hourInterval = $dateTime->format('YmdH');
+
+        if (Config::get('database.default') === 'mysql') {
+            $points = $metric->points()
+                                ->whereRaw('DATE_FORMAT(created_at, "%Y%m%d%H") = '.$hourInterval)
```
Comment (GrahamCampbell) [inline app/Repositories/MetricRepository.php]: indentation

File: app/Repositories/MetricRepository.php [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+     * @return int
+     */
+    protected function getPointsForDayInWeek(Metric $metric, $day)
+    {
+        $dateTime = (new Date())->setTimezone($this->dateTimeZone);
+        $dateTime->sub(new DateInterval('P'.$day.'D'));
+
+        if (Config::get('database.default') === 'mysql') {
+            $points = $metric->points()
+                                ->whereRaw('created_at BETWEEN DATE_SUB(created_at, INTERVAL 1 WEEK) AND NOW()')
```
Comment (GrahamCampbell) [inline app/Repositories/MetricRepository.php]: same

File: app/Repositories/MetricRepository.php (hunk 2 above)
Comment (jbrooksuk) [inline app/Repositories/MetricRepository.php]: Outdent one?

File: app/Http/Controllers/HomeController.php [hunk 4]
```diff
@@ -52,9 +70,17 @@ public function showIndex()
...
-
+        $metricData = [];
         if ($displayMetrics = Setting::get('display_graphs')) {
             $metrics = Metric::where('display_chart', 1)->get();
+
+            $metrics->map(function ($metric) use (&$metricData) {
+                $metricData[$metric->id] = [
+                    'today'  => $this->metricRepository->listPointsToday($metric),
+                    'week'   => $this->metricRepository->listPointsForWeek($metric),
+                    'month'  => $this->metricRepository->listPointsForMonth($metric),
```
Comment (GrahamCampbell) [inline app/Http/Controllers/HomeController.php]: indentation

Comment (GrahamCampbell) [inline app/Http/Controllers/HomeController.php]: alignment

File: app/Repositories/MetricRepository.php (hunk 2 above)
Comment (GrahamCampbell) [inline app/Repositories/MetricRepository.php]: should only be indented by 4 more spaces than the points variable

File: app/Http/Controllers/HomeController.php (hunk 4 above)
Comment (jbrooksuk) [inline app/Http/Controllers/HomeController.php]: What's up with the alignment and indentation here? Looks fine?

**Comment by GrahamCampbell** [comment]
The composer.lock file shouldn't be updated here.

**Comment by jbrooksuk** [comment]
Whoops. I can't remember why I did that.

File: app/Http/Controllers/HomeController.php (hunk 4 above)
Comment (GrahamCampbell) [inline app/Http/Controllers/HomeController.php]: nope

Comment (GrahamCampbell) [inline app/Http/Controllers/HomeController.php]: one too many spaces

Comment (jbrooksuk) [inline app/Http/Controllers/HomeController.php]: Oh

**Pull request merged**